cq/cqlearners.py

```diff
@@ -0,0 +1,196 @@
+        self.q_values[state][action] = old_q_value + self.learning_rate * (new_q_value - old_q_value)
+        
+        
+class GlobalActionSelector(ActionSelector): 
+    def setup(self):
+        self.q_values = {}
+    
+    def addState(self, state):
+        if (state not in self.q_values.keys()):
+            self.q_values[state] = np.zeros(self.no_of_actions)
+        
+    def updateQValuesWithMaxQ(self, state, action, reward, maxQValue): 
+        if (state not in self.q_values.keys()):
+            self.q_values[state] = np.zeros(self.no_of_actions)
+            
+        old_q_value = self.q_values[state][action]
+        new_q_value = reward + self.discount_factor * maxQValue
+        self.q_values[state][action] = old_q_value + self.learning_rate * (new_q_value - old_q_value)
+        
+        
+        
+
+class CQLearner: 
+    
+    coordination_joint_states = None
+    coordination_joint_states_confidence = None
+    
+    local_action_selector = None
+    global_action_selector = None
+    
+    initial_rewards = None
+    latest_rewards = None
+    
+    no_of_actions = None
+    no_of_states = None
+    
+    epison = None
+    discount_factor = None
+    learning_rate = None
+    sliding_window_size = None
+    name = None
+    
+    def __init__(self, name, no_of_states, no_of_actions, sliding_window_size=60, epison=0.1, discount_factor=0.9, learning_rate=0.1):
+        self.name = name
+        self.no_of_actions = no_of_actions
+        self.no_of_states = no_of_states
+        self.epison = epison
+        self.discount_factor = discount_factor
+        self.learning_rate = learning_rate
+        self.sliding_window_size = sliding_window_size
+        
+        self.coordination_joint_states = []
+        self.coordination_joint_states_confidence = {}
+        
+        self.local_action_selector = LocalActionSelector(name, no_of_states, no_of_actions, epison, discount_factor, learning_rate)
+        self.local_action_selector.setup()
+        self.global_action_selector = GlobalActionSelector(name, no_of_states, no_of_actions, epison, discount_factor, learning_rate)
+        self.global_action_selector.setup()
+    
+        self.initial_rewards = [ [deque(maxlen=sliding_window_size) for a in range(no_of_actions)] for s in range(no_of_states) ]
+        self.latest_rewards = [ [deque(maxlen=sliding_window_size) for a in range(no_of_actions)] for s in range(no_of_states) ]
+    
+    
+    def reset(self):
+        self.coordination_joint_states = []
+        self.coordination_joint_states_confidence = {}
+        #self.latest_rewards = [ [deque(maxlen=self.sliding_window_size) for a in range(self.no_of_actions)] for s in range(self.no_of_states) ]
+    
+    def takeAction(self, local_state, global_state):
+        isNecessary, joint_states_including_local_state = self.__isCoordinationNecessary(local_state)
+
+        if (isNecessary and (global_state in joint_states_including_local_state) 
```

Comment:
In this case, I think that `global_state in joint_states_including_local_state` will already imply `isNecessary`. Maybe we only need to check `global_state in joint_states_including_local_state` then?

If I understand correctly - 
* `isNecessary`: there is at least some joint state in `coordination_joint_states` where our locally state is present
* `global_state in joint_states_including_local_state`: there is some state in `coordination_joint_states` that exactly matches the global state

Reply:
Your interpretation is correct, but it's not the same thing. 

Actually, the 2 validations should be implemented into 2 steps since the agent has no clue about global state when he does the first condition, but it's still okay to implement it that way. 

It goes as follow: 
1. agent checks isNecessary based on information from his local state
2. if yes (there is a potential that he in nonsafe state), it then notice the global state and make sure it really exists in joint_states_including_local_state
3. if it exists in both, this mean that this global joint state is really unsafe
4. if not, it means that it is not 


for example, agent1 entered state 1 and agent 2 entered state 2
agent1 detects conflict and added (1,2) to his unsafe states. 
after a while, it entered state 1 again while agent2 entered state 9
agent1 will see that a conflict occurred before in that state
he checks global state (1,9) which is not in his unsafe states, 
then he just use his local q values and decrease confidence for (1,2)
on the other hand, if agent2 was in the same state 2
agent1 will find that (1,2) is actually in his list, it will use its global q values and increase confidence for (1,2)
and so on..

it is somehow a mechanism to determine whether the other agent is close to me or not

Reply:
I agree - I don't think it's a bug. Just trying to make sure I understand the code / algorithm.

Reply:
Thanks for the response clarifying :)